src/Persistence/Search/FacetSearchIndexFieldsBuilder.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace ProfessionalWiki\WikibaseFacetedSearch\Persistence\Search;
+
+use Exception;
+use ProfessionalWiki\WikibaseFacetedSearch\Application\Config;
+use SearchEngine;
+use SearchIndexField;
+use Wikibase\DataModel\Services\Lookup\PropertyDataTypeLookup;
+
+class FacetSearchIndexFieldsBuilder {
+
+	public function __construct(
+		private readonly SearchEngine $engine,
+		private readonly Config $config,
+		private readonly PropertyDataTypeLookup $dataTypeLookup
+	) {
+	}
+
+	/**
+	 * @return array<string, SearchIndexField>
+	 */
+	public function createFields(): array {
+		$fields = [];
+
+		foreach ( $this->config->getFacets()->asArray() as $facetConfig ) {
+			$name = 'wbfs_' . $facetConfig->propertyId->getSerialization();
+
+			try {
+				$dataTypeId = $this->dataTypeLookup->getDataTypeIdForProperty( $facetConfig->propertyId );
+			} catch ( Exception $e ) {
+				continue;
+			}
+
+			$fieldType = $this->getFieldTypeForDataTypeId( $dataTypeId );
+
+			if ( $fieldType === null ) {
+				continue;
+			}
+
+			$fields[$name] = $this->engine->makeSearchFieldMapping( $name, $fieldType );
```

Comment:
`$name` can be computed closer to where it is used and the details of said computation could go into a dedicated private method.